### ISSUE TITLE
Realm crashes for a long running program

### ISSUE THREAD

**Issue by magnatelee** [StanfordLegion/legion#336]
Realm raises an assertion when I run circuit for 500k iterations with Legion Prof. Here is a command to reproduce:

`regent.py examples/circuit_sparse.rg -l 500000 -lg:prof_logfile prof_% -lg:prof 1 `

After running for a while, it will raise the following assertion:

`realm/profiling.inl:244: void Realm::ProfilingMeasurementCollection::add_measurement(const T&, bool) [with T = Realm::ProfilingMeasurements::InstanceAllocResult]: Assertion 'measurements.count(static_cast<ProfilingMeasurementID>(T::ID)) == 0' failed.`

A cursory look at the stack trace seems to suggest that the profiling measurements in a region instance are not cleared when the instance gets reused at the Realm level:

```
(gdb) bt
...
#8  0x00007f4620a6be65 in Realm::ProfilingMeasurementCollection::add_measurement<Realm::ProfilingMeasurements::InstanceAllocResult> (this=0x7f45e4617848, data=..., send_complete_responses=true) at /home/wclee/le
gion/runtime/realm/profiling.inl:244
#9  0x00007f4620a69d38 in Realm::RegionInstanceImpl::notify_allocation (this=0x7f45e4617820, success=true, offset=4672) at /home/wclee/legion/runtime/realm/inst_impl.cc:480
#10 0x00007f4620a59f19 in Realm::MemoryImpl::allocate_instance_storage (this=0x494b670, i=..., bytes=12, alignment=32, precondition=...) at /home/wclee/legion/runtime/realm/mem_impl.cc:412
#11 0x00007f4620a6842b in Realm::RegionInstance::create_instance (inst=..., memory=..., ilg=0x7f45a1ad03c0, prs=..., wait_on=...) at /home/wclee/legion/runtime/realm/inst_impl.cc:109
...
(gdb) f 8
#8  0x00007f4620a6be65 in Realm::ProfilingMeasurementCollection::add_measurement<Realm::ProfilingMeasurements::InstanceAllocResult> (this=0x7f45e4617848, data=..., send_complete_responses=true) at /home/wclee/legion/runtime/realm/profiling.inl:244
244         assert(measurements.count(static_cast<ProfilingMeasurementID>(T::ID)) == 0);
(gdb) p measurements
$5 = std::map with 2 elements = {
  [Realm::PMID_INST_ALLOCRESULT] = {
    <Realm::ByteArrayRef> = {
      array_base = 0x7f45e4618110, 
      array_size = 1
    }, <No data fields>},
  [Realm::PMID_INST_MEM_USAGE] = {
    <Realm::ByteArrayRef> = {
      array_base = 0x7f45e46183d0, 
      array_size = 24
    }, <No data fields>}
}
(gdb) p T::ID
$6 = Realm::PMID_INST_ALLOCRESULT
```

**Comment by lightsighter** [comment]
@streichler is on vacation, but I think I know what happened. Currently Realm is not recycling RegionInstanceImpl meta-data objects, and interestingly the ID of the RegionInstance that we failed on here is:
(gdb) p/x i
$4 = {
  id = 0x6000000000000000
}
So what happened here is in the process of making and destroying (but not recycling) a bunch of reduction instances, we consumed all the bits available for describing physical instances in this memory and we then wrapped around back to 0, so this is the first time we tried to reuse RegionInstanceImpl, which of course hadn't had 'RegionInstanceImpl::finalize_instance' called on it, so there were still profiling requests in it.

Long term @streichler should fix this in a better way, but if in the short term we need to run this for longer I can show @magnatelee how to increase the number of bit used for describing physical instances in a memory so we can run for longer.

**Comment by elliottslaughter** [comment]
By way, we also hit this with psana if we run over 32k events. So while not urgent (yet), we should consider this a bug that will at some point impact a production code.

**Comment by streichler** [comment]
@magnatelee @elliottslaughter This is hopefully finally fixed with commit f6aa30e.  Please try your various test cases to confirm/refute my claim.

**Comment by magnatelee** [comment]
I was able to run Circuit for 30 minutes without any crash, so I think this issue has been fixed. That said, the memory usage reached 36 GB after 30 minutes even though the size of circuit is only 4 KB. This may or may not be related to the fix and I feel we should make a separate GitHub issue to chase this. @elliottslaughter Feel free to close the issue if you confirm it has been fixed in psana

**Comment by streichler** [comment]
Yes, let's open a separate issue to track the memory consumption.  I just tried running my Realm-only instance-reuse unit test and it shows no memory leaks, so I'd suggest we start with having Mike look at his allocation-tracking counters at the Legion level.

**Comment by lightsighter** [comment]
@magnatelee I think you already know how to use the -DTRACE_ALLOCATION compile-time flag?

**Comment by magnatelee** [comment]
@lightsighter Sure, but I think we don't need to use that because I'm pretty sure the leak is in Legion Prof. The memory usage stays relatively constant without Legion Prof. Let me see if I can quickly figure this out.

**Comment by lightsighter** [comment]
@magnatelee and I figured this out. It is due to the profiling responses being starved from running in Realm due to the low priority that we give them to avoid interfering with the actual execution. I made a change which might allow them to find more "holes" in which to run including holes on the actual application processors. We also could potentially change their priority to avoid starving them but at the cost that the profiler might interfere more with the application performance. Right now I think since we mostly do profiler captures for shorter runs it is better to avoid interfering with the application performance and pay a small memory price for starvation of profiling responses until the end of the run when we can process them. @magnatelee can we close this issue now?
